### PR TITLE
Added initial interface for hypervolume contribution algorithms

### PR DESCRIPTION
**Pull request by Ulfgard**
This pull request is mostly there to start an initial code review and discussion on the design on an interface for the contribution algorithms. 

I went for the route which returns the indices of the k-least or k largest contributors with and without reference point.

I have added a 3D-Version of the algorithm.

### CODE REVIEW

File: include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		std::vector<std::pair<double,std::size_t> > bestK(k+1);
+		auto heapStart = bestK.begin();
+		auto heapEnd = bestK.begin();
+		
+		auto pointComp = [&](std::pair<double,std::size_t> const& lhs, std::pair<double,std::size_t> const& rhs){return comp(lhs.first,rhs.first);};
+		
+		//compute the hypervalue contribution for each point except the endpoints
+		*heapEnd = std::make_pair((front[2].f1 - front[1].f1) * (front[0].f2 - front[1].f2),front[1].index);
+		++heapEnd;
+		for(std::size_t i = 2; i < front.size()-1;++i){
```
Comment (TGlas) [inline include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h]: Hmmm... if you want all but the extremes, shouldn't the loop start at i = 1 ?

File: include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		
+		return bestContributors(front,k,[](double con1, double con2){return con1 < con2;});
+	}
+	
+	/// \brief Returns the index of the points with largest contribution.
+	///
+	/// \param [in] points The set \f$S\f$ of points from which to select the least contributor.
+	/// \param [in] referencePoint The reference point \f$\vec{r} \in \mathbb{R}^2\f$ for the hypervolume calculation, needs to fulfill: \f$ \forall s \in S: s \preceq \vec{r}\f$.
+	template<typename Set, typename VectorType>
+	std::vector<std::size_t> largest( Set const& points, std::size_t k, VectorType const& referencePoint)const{
```
Comment (TGlas) [inline include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h]: "largest" is the opposite of "smallest", "most" is the opposite of "least". Another possible pair would be worst/best, which is probably closest to what we want here. So I propose to rename least -> worst and largest -> best.

**Comment by TGlas** [comment]
Where exactly do you see a need for discussing the interface design? It is probably not just the names, right?

File: include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h (hunk 1 above)
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h]: no, the point i=1 is already inserted in line 82. hm maybe i can unify this...yeah probably

**Comment by Ulfgard** [comment]
for example: do we also want the contribution itself?

**Comment by TGlas** [comment]
> for example: do we also want the contribution itself?

This is for selection in an EA. Right now I don't foresee any other 
potential applications (at least, within Shark). For this purpose the 
contribution does not matter much.

However, in other contexts the code could be used for performance 
monitoring and analysis (e.g., in BBOB). Then the size of the 
contribution may matter. Since it is anyway available, why not make it 
accessible?

**Comment by Ulfgard** [comment]
is this now okay?

**Comment by TGlas** [comment]
Yes, thanks, great!

**Pull request merged**